### PR TITLE
Normalize menu links for user/* paths, so that URLs containing uids a…

### PR DESCRIPTION
**Pull request by gielfeldt**
…rent cached.

### CODE REVIEW

File: ding_user_frontend.module [hunk 1]
```diff
@@ -330,3 +330,31 @@ function ding_user_frontend_build_menu($path = NULL) {
...
+  return preg_replace('/^user\/(.*?\/)(.*)$/', 'user/me/$2', $path);
+}
+
+/**
+ * Implemengs hook_url_inbound_alter().
+ *
+ * Re-replace the "me" in user/* paths with the proper user id.
+ */
+function ding_user_frontend_url_inbound_alter(&$path, $original_path, $path_language) {
+  if (preg_match('/^user\/me\/(.*)$/', $path, $matches)) {
```
Comment (kasperg) [inline ding_user_frontend.module]: There seems to be little use for regular expressions here. Would it be more efficient to do simple string comparison between the first 8 chars of the path and `user/me/`?

File: ding_user_frontend.module [hunk 2]
```diff
@@ -330,3 +330,31 @@ function ding_user_frontend_build_menu($path = NULL) {
...
   return $menu;
 }
+
+/**
+ * Build a consistent link for user/* paths.
+ *
+ * If URLs like user/1234/view are cached in e.g. varnish, then other users
+ * may be exposed to these URLs.
+ *
+ * This function normalizes the user/* paths, by replacing the uid with "me".
```
Comment (kasperg) [inline ding_user_frontend.module]: :+1:

File: ding_user_frontend.module [hunk 3]
```diff
@@ -330,3 +330,31 @@ function ding_user_frontend_build_menu($path = NULL) {
...
+ *
+ * @param  [type] $path [description]
+ * @return [type]       [description]
+ */
+function ding_user_frontend_build_uniform_menu_link($path) {
+  return preg_replace('/^user\/(.*?\/)(.*)$/', 'user/me/$2', $path);
+}
+
+/**
+ * Implemengs hook_url_inbound_alter().
```
Comment (kasperg) [inline ding_user_frontend.module]: Spelling error :)

File: ding_user_frontend.module [hunk 4]
```diff
@@ -330,3 +330,31 @@ function ding_user_frontend_build_menu($path = NULL) {
...
+ *
+ * Re-replace the "me" in user/* paths with the proper user id.
```
Comment (kasperg) [inline ding_user_frontend.module]: An explicit reference to `ding_user_frontend_build_uniform_menu_link` might be useful here.

**Comment by kasperg** [comment]
@gielfeldt Review complete. Very well done. Avoiding `hook_url_outbound_alter()` is a good thing.

File: ding_user_frontend.module (hunk 1 above)
Comment (gielfeldt) [inline ding_user_frontend.module]: `hook_url_inbound_alter` is called once per request. I believe the performance impact to be negligible.

Comment (kasperg) [inline ding_user_frontend.module]: OK.

**Comment by gielfeldt** [comment]
@kasperg Updated according to review.

**Comment by kasperg** [comment]
Approved.

**Pull request closed without merging**